src/Blocks/BlockWallSign.h

```diff
@@ -33,39 +33,41 @@ class cBlockWallSignHandler:
 
 
 
-	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, int a_RelX, int a_RelY, int a_RelZ, const cChunk & a_Chunk) override
+	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		int BlockX = (a_Chunk.GetPosX() * cChunkDef::Width) + a_RelX;
-		int BlockZ = (a_Chunk.GetPosZ() * cChunkDef::Width) + a_RelZ;
-		GetBlockCoordsBehindTheSign(a_Chunk.GetMeta(a_RelX, a_RelY, a_RelZ), BlockX, BlockZ);
-		BLOCKTYPE Type = a_ChunkInterface.GetBlock({BlockX, a_RelY, BlockZ});
-
-		return ((Type == E_BLOCK_WALLSIGN) || (Type == E_BLOCK_SIGN_POST) || cBlockInfo::IsSolid(Type));
+		auto NeighborPos = a_RelPos + GetOffsetBehindTheSign(a_Chunk.GetMeta(a_RelPos));
+		BLOCKTYPE NeighborType;
+		a_Chunk.UnboundedRelGetBlockType(NeighborPos, NeighborType);
```

Comment:
to be fair I think we've somewhat thrown in the towel with the whole decoupling thing (vis a vis #3162)

Reply:
Because `cChunkInterface` is the interface provided by `cWorld`, while I'd like to eventually move all this processing to being relevant to `cChunk` (or an interfaced version of it), so that the accesses don't have to go through the chunkmap.

Honestly, I find these Interface classes somewhat unfortunate, they don't reflect the usage we want from them, they instead reflect the API provided by what they started from. A better interface would be "Block type accessor", whom you could simply ask to get you block type at specific coords; it could be provided by a cWorld, cChunk or even cBlockArea.

Unless you insist on going back with this, I'd prefer to keep things as they are here until a bigger refactoring.

Reply:
I feel quite strongly that we should avoid `cChunk`. `cChunk` is more complicated to use. You need to think about:
- absolute vs. relative coordinates
- `Get` vs. `UnboundedRelGet`
- `UnboundedRelGet` can also fail (which isn't handled here)

>so that the accesses don't have to go through the chunkmap.

And what specifically is wrong with that? Are you concerned about performance?

>Unless you insist on going back with this, I'd prefer to keep things as they are here until a bigger refactoring.

It was using `cChunkInterface` before which is why I commented on the change. You're actively choosing to rewrite it this way.

Reply:
The original code still needs to use `a_Chunk` at least for the coord conversion, the new code uses one object less.

Performance is only a part of my concern, but yes, the chunkmap needs to search for the right chunk for every access, and `CanBeAt` can be at code paths that are quite hot.

`cChunkInterface` can also fail to locate the block (chunk unloaded), and it fails silently without ever letting us know, so the behavior is then incorrect for those cases.

Reply:
And to counter your list, I actually think `cChunk` is easier to use:
 - there are only relative coords (unlike `cBlockArea`), and the handler already gets those
 - `Get` is for same-XZ coords, `Unbounded` is for when moving to neighbors. Simple.
 - `UnboundedRelGet` will notify of the failure; `cChunkInterface` just returns bogus data (`0`) on failure, with no way of detecting it.

Reply:
>Performance is only a part of my concern, but yes, the chunkmap needs to search for the right chunk for every access

When processing a chunk, the chunk map entry is going to be hot in cache so I have doubts that this is a significant effect. Hard to say without some performance data.

>cChunkInterface can also fail to locate the block (chunk unloaded), and it fails silently without ever letting us know, so the behavior is then incorrect for those cases.

That's a good point. The future `cBlockTypeInterface` (or whatever) should probably return `optional<BLOCKTYPE>`.

I suggest we move this discussion to an issue so it's not blocking this PR.